staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -496,6 +509,32 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		chanSize = 1000
 	}
 
+	// We explicitly use thread unsafe version and do locking ourself to ensure that
+	// no new events will be processed in the meantime. The watchCache will be unlocked
+	// on return from this function.
+	// Note that we cannot do it under Cacher lock, to avoid a deadlock, since the
+	// underlying watchCache is calling processEvent under its lock.
+	c.watchCache.RLock()
+	defer c.watchCache.RUnlock()
+
+	// Determine if a consistent stream was requested
+	bookmarkAfterResourceVersion := uint64(0)
+	if opts.SendInitialEvents && utilfeature.DefaultFeatureGate.Enabled(features.WatchList) {
+		switch {
+		case len(opts.ResourceVersion) == 0:
+			rv, err := c.getCurrentResourceVersionFromStorage(ctx)
+			if err != nil {
+				return newErrWatcher(err), nil
+			}
+			bookmarkAfterResourceVersion = rv
+		case watchRV == 0:
+			// read the current RV under the lock
+			bookmarkAfterResourceVersion = c.watchCache.resourceVersion
+		default:
+			bookmarkAfterResourceVersion = watchRV
```

Comment:
alternatively, we could check if the server has an RV >= the desired one and stop if it doesn't. In that case, we would have to close the current connection with the client so it could hit a different replica.